yaas/cluster.py

```diff
@@ -1,4 +1,182 @@
 # coding: utf-8
+from __future__ import absolute_import
+from __future__ import print_function
+import argparse
+import inspect
+import json
+import pprint
+import requests
+import sys
+
+from . import config
 
 def command(parser, args):
-    pass
+    """Create, modify, and view clusters"""
+    subparsers = parser.add_subparsers(dest="command")
+
+    commands = {
+        'list': _list,
+        'create': _create,
+        'show': _show,
+        'blueprint': _blueprint,
+        'destroy': _destroy,
+        'start': _start,
+        'stop': _stop,
+        }
+
+    for name, fn in commands.iteritems():
+      subparsers.add_parser(name, help=inspect.getdoc(fn))
+
+    sub_args, extra = parser.parse_known_args(args)
+    commands[sub_args.command](subparsers.choices[sub_args.command], extra)
+
+def _list(parser, args):
+    """ List all clusters. """
+    parser.add_argument(
+        '-f',
+        '--fields',
+        help="Print host details.")
+    list_args, extra = parser.parse_known_args(args)
+    response = requests.get(
+        config.href('/api/v1/clusters'),
+        params={'fields': list_args.fields},
+        **config.requests_opts())
+    if config.args.raw:
+        pprint.pprint(response.json())
+        sys.exit(0)
+    response.raise_for_status()
+    res = response.json()
+    for item in res['items']:
+        line = []
+        for key in ('cluster_name', 'version'):
+            if key in item['Clusters']:
+                line.append(item['Clusters'][key])
+        print(' - '.join(line))
+        if list_args.fields:
+            for key, value in item.items():
+                config.print_field(k=key, v=value, indent=4)
+
+def _create(parser, args):
+    """Create a cluster from a blueprint"""
+    parser.add_argument(
+        'name',
+        help="Cluster name")
+    parser.add_argument(
+        '-t',
+        '--template',
+        nargs='?',
+        type=argparse.FileType('r'),
```

Comment:
This isn't about premature optimization; it's about correctness (i.e. "It's broken, fix it" lol). That's why there's a bug open about it. We shouldn't knowingly put any kind of leak into anything we write. Also, I find the bug author's first comment relevant here:

> This is well documented as an anit-idiom in http://docs.python.org/howto/doanddont.html#exceptions.

If this bug has been fixed in Python 3, I can get behind this approach. Either way, this probably needs to change to support creating clusters without a template (assuming it's possible... looking into that shortly).